Fix has_many_inversing with touch: true on a belongs_to association with inverse

doits opened the pull request:
### Summary

`has_many_inversing` adds records to a has_many association. It does so with destroyed records, too. So if a child was destroyed with a `belongs_to :parent, touch: true` association *and* the parent was not loaded, it tried to load the parent to touch it. While loading the parent it added the child record to the parent's has_many association. The logic doing this always set the child's parent id – even if it was correct/the same already. But since the child is destroyed, it resulted in a `FrozenError`.

This commit prevents doing the unnecessary setting of the identical id and therefore fixes this error.

Fixes #40943

doits committed:
Fix has_many_inversing with touch: true on a belongs_to association with inverse

`has_many_inversing` adds records to a has_many association. It does so
with destroyed records, too. So if a child was destroyed with a `touch:
true` association on the parent *and* the parent was not loaded, it
tried to load the parent to touch it. While loading the parent it added
the child record to the parent's has_many association. The logic doing
this always set the child's parent id – even if it was correct/the same
already. But since the child is destroyed, it resulted in a
`FrozenError`.

This commit prevents doing the unnecessary setting of the identical id
and therefore fixes this error.

Fixes #40943 by not doing an unneeded attribute set.

rafaelfranca commented on activerecord/lib/active_record/associations/belongs_to_association.rb:
```diff
@@ -29,7 +29,7 @@ def handle_dependency
 
       def inversed_from(record)
-        replace_keys(record)
+        replace_keys(record) if owner[reflection.foreign_key] != record_target_key(record)
```
This will now read the attribute twice. Since `replace_key` is internal we should pass the key as second argument to avoid the extra lookup.

doits replied:
I think you mean the call to `record_target_key` right? I changed it so it is only called once, albeit I'm not sure if there is a better default for the second parameter of `replace_keys`.

doits committed:
cache target key so it is read only once

rafaelfranca commented on activerecord/lib/active_record/associations/belongs_to_association.rb:
```diff
@@ -108,8 +109,12 @@ def require_counter_update?
 
-        def replace_keys(record)
-          owner[reflection.foreign_key] = record ? record._read_attribute(primary_key(record.class)) : nil
+        def replace_keys(record, target_key = :unset)
```
We can alway pass the second argument to all method calls. It doesn't need to be optional. There is only two method calls for this method and all of them will need to calculate `record_target_key`.

doits replied:
:+1: The failing specs showed that the polymorphic case needs an update then, too. I had to refactor it a little bit, let's see if the tests pass 🤞

rafaelfranca replied:
I think you went too much. It is better to always pass two parameters and only update the polymorphic override of `replace_keys` to accept two arguments instead of one.

doits replied:
OK, updated it, but now I had to override `replace`, too, since it calls `replace_keys` with only one parameter in the non polymorphic case, so there is some code duplication. See 551824adc82e04edef922167bba496e47049780d.

I then refactored it more to make it more dry in a60ee1414b415c35bbed50698e8edc017443c1e8 by introducing a `replace_keys_if_needed` method – what do you say about this?

BTW: I also tried to change `replace_keys` to *always* only set it if needed (e.g. if different), but this broke an other test that expects a frozen error setting an association on a destroyed record directly.

rafaelfranca replied:
I tried a simpler version of this patch and all tests passed. Don't this diff works in top of your branch?

```diff
diff --git a/activerecord/lib/active_record/associations/belongs_to_association.rb b/activerecord/lib/active_record/associations/belongs_to_association.rb
index b7ce4ca8d3..b4c2bc9dfd 100644
--- a/activerecord/lib/active_record/associations/belongs_to_association.rb
+++ b/activerecord/lib/active_record/associations/belongs_to_association.rb
@@ -29,7 +29,8 @@ def handle_dependency
       end
 
       def inversed_from(record)
-        replace_keys_if_needed(record)
+        target_key = record_target_key(record)
+        replace_keys(record, target_key) if owner[reflection.foreign_key] != target_key
         super
       end
 
@@ -80,7 +81,7 @@ def replace(record)
             @updated = true
           end
 
-          replace_keys(record)
+          replace_keys(record, record_target_key(record))
 
           self.target = record
         end
@@ -108,13 +109,8 @@ def require_counter_update?
           reflection.counter_cache_column && owner.persisted?
         end
 
-        def replace_keys(record)
-          owner[reflection.foreign_key] = record_target_key(record)
-        end
-
-        def replace_keys_if_needed(record)
-          target_key = record_target_key(record)
-          owner[reflection.foreign_key] = target_key if owner[reflection.foreign_key] != target_key
+        def replace_keys(record, target_key)
+          owner[reflection.foreign_key] = target_key
         end
 
         def record_target_key(record)
diff --git a/activerecord/lib/active_record/associations/belongs_to_polymorphic_association.rb b/activerecord/lib/active_record/associations/belongs_to_polymorphic_association.rb
index 9bf492962e..ff428abf3d 100644
--- a/activerecord/lib/active_record/associations/belongs_to_polymorphic_association.rb
+++ b/activerecord/lib/active_record/associations/belongs_to_polymorphic_association.rb
@@ -14,20 +14,9 @@ def target_changed?
       end
 
       private
-        def replace_keys(record)
+        def replace_keys(record, _target_key)
           super
-          owner[reflection.foreign_type] = record_target_type(record)
-        end
-
-        def replace_keys_if_needed(record)
-          super
-
-          target_type = record_target_type(record)
-          owner[reflection.foreign_type] = target_type if owner[reflection.foreign_type] != target_type
-        end
-
-        def record_target_type(record)
-          record ? record.class.polymorphic_name : nil
+          owner[reflection.foreign_type] = record ? record.class.polymorphic_name : nil
         end
 
         def inverse_reflection_for(record)
```

doits replied:
Yeah the tests pass indeed, and it is much simpler. Though it handles one thing differently in the polymorphic case, but maybe that is no problem:

`inversed_from` now calls `replace_keys` only if the `target_key` changed. This means in the *polymorphic* association it will *not* call `replace_keys` if *only* the `type` changed, but the `target_key` did *not* – so the different `type` will *not* be set in such a case. (In my longer version it sets the different `type` regardless if the `target_key` is the same.)

But can this happen? Can there ever be a call of `inversed_from` with a `record` that has a *different* `type` but the *same* `target_key` as already set?

Right now this is over my head (and I need some sleep). Do you know if this could happen or can you maybe rule it out?

doits replied:
I commited the simpler version so the latest version is in this PR. I'm still not 100 % sure if the edge case in the polymorphic association – same id but different type in `inversed_from` – can happen (I'd probably keep the check just to be safe) but I'll leave that to you to decide.

Oh and happy new year btw ;-)

doits committed:
refactor and fix for polymorphic associations, too

doits committed:
no need for record param now

doits committed:
amend after feedback

doits committed:
refactor to make it more dry

doits committed:
commit simpler version

it does not check the type in `inversed_from` in the polymorphic case
but that is probably not necessary

doits commented:
👍 gj, actually had a `force` param in my mind at some point, too - not sure why I didn't implented it. I think this is the best solution now.
